OpenBadges Generation

Kelketek opened the pull request:
**Background**: This is the first PR for the OpenBadges MVP. It contains only the backend code for generating the certificates based on a post-save hook on the Certificate model.
**Jira ticket number**: https://openedx.atlassian.net/browse/SOL-768
Previous discussions: [OpenBadges Spec](https://docs.google.com/document/d/1oLIHSfeRZw1-s8C9zHucvVnqfsHnvh3Vq0inu5DyZ5k/edit?usp=sharing)
**Partner information**: 3rd party-hosted open edX instance, for an edX solutions client.

This PR requires setting up a sandbox with certificate generation as well as access to the command line to fully test, as a UI is not yet present for this functionality. Additional PRs with a UI will be based atop this one, and this one is provided primarily for code review. It may be best if this PR is merged into a feature branch, with the final PR being that branch against master once the full circuit of functionality is reviewable on a sandbox.

@smarnach This is ready for your review.

edx-webhook commented:
Thanks for the pull request, @Kelketek! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=7888&repo=edx%2Fedx-platform

e-kolpakov commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,68 @@
+        self.base_url = "{}/v1/issuer/issuers/{}".format(settings.BADGR_BASE_URL, self.issuer)
+        self.badge_create_url = "{}/badges".format(self.base_url)
+        self.badge_url = "{}/{}".format(self.badge_create_url, self.course_slug)
+        self.assertion_url = "{}/assertions".format(self.badge_url)
```
@Kelketek nit: these four urls could be properties or even lazy properties

e-kolpakov commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,68 @@
+    def get_headers(self):
+        return {'Authorization': 'Token {}'.format(self.token)}
+
+    def verify_badge(self):
```
@Kelketek nit: I would expect `verify_*` to check some condition and return True/False. Looks like this method ensures the badge is created, so it might make sense to give him new name to reflect that, i.e. `ensure_badge_created`

e-kolpakov commented on lms/djangoapps/certificates/models.py:
```diff
@@ -547,3 +549,39 @@ def get_config(cls):
+    if not modulestore().get_course(instance.course_id).issue_badges:
+        return
+    if BadgeTracker.objects.filter(user=instance.user, course_id=instance.course_id):
+        # Badge already exists. Skip.
```
@Kelketek might make sense to add some logging here and on line 574: these are special situations so it might help debugging it someday.

edx-webhook commented:
Thanks for the pull request, @Kelketek! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=7888&repo=edx%2Fedx-platform

smarnach commented:
Please remove `common/test/data/badges/._bulb.svg`.

smarnach commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,134 @@
+    """
+    The only properly public method of this class is 'award'. If an alternative object is created for a different
+    badging service, the other methods don't need to be reproduced.
+    """
```
Don't we follow PEP 257 for docstring formatting?  If not, what style guide applies?

Kelketek replied:
We don't have a particular style guide for docstrings. We require you to have them and that pylint doesn't complain about the line length. I've seen quite a bit of variance on this but it hasn't seemed to be something that upstream has wanted to make an issue of, in the scheme of everything else that needs to be focused on.

The most important thing is that is has docstrings to begin with, after all. :)

smarnach replied:
Fair enough, good to know.  I was used to quite strict rules for docstrings (and actually found them quite useful, to my surprise).

lamagnifica replied:
FYI there is some info about formatting docstrings in edX code in our wiki here: https://openedx.atlassian.net/wiki/display/AC/edX+REST+API+Conventions#edXRESTAPIConventions-docstringsDocstringFormat

smarnach commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,134 @@
+from django.conf import settings
+from django.contrib.sites.models import Site
+from django.core.urlresolvers import reverse
+from lazy.lazy import lazy
```
Just as an observation, I usually did `from lazy import lazy`.  It turns out this is completely equivalent, but being lazy, I'll stick with the shorter form. :)

Kelketek replied:
Even lazier, I just type lazy somewhere in the code, right-click, tell IntelliJ to import, and this is what it comes up with :P Good to know, though.

smarnach commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,134 @@
+        image_name = settings.BADGR_IMAGE_SOURCES[mode]
+        image = open(image_name, 'rb')
+        # This will fail if the filename is something silly.
+        content_type = 'image/{}'.format(image_name.rsplit('.', 1)[-1].lower())
```
How about using `os.path.splitext(filename)[1][1:]` to at least avoid throwing an exception if the file name is silly?

Kelketek replied:
@smarnach Could you give me an example of a string that would throw an exception here? I thought I had constructed this to where it wouldn't.

I mean, it may end up constructing a completely nonsense content-type, but it should still upload the file. At which point we'll get an exception either way when the target server complains.

smarnach replied:
You are right, it doesn't throw.  I still generally prefer to use a library method that does exactly what I need instead of rewriting it.  `splitext()` immediately tells what it does, so the code becomes easier to read, and using library functions is generally less error-prone than ad-hoc rewrites of the same functionality.

Kelketek replied:
Misread the method name at first. I agree, that would be better. I'll swap em out :)

smarnach commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,134 @@
+        about_path = reverse('about_course', kwargs={'course_id': unicode(self.course_key)})
+        data = {
+            'name': course.display_name,
+            'criteria': u'http://{}{}'.format(domain, about_path),
```
Do we really want to hard-code the protocol here?  The server might be HTTPS-only.  I don't see an easy way of figuring this out without having a request available, but just a thought.

Kelketek replied:
@smarnach I don't really see a way around it. The more troubling thought to me is if the Site object is maintained at all on individual installations, or if there's a better way to get the LMS domain through the LMS.

Any competent deployment engineer should have redirects set up if this needs to be turned into HTTPS anyway, though I agree it's not a pleasant situation.

antoviaque replied:
See https://github.com/edx/edx-platform/blob/d4de932c2b46dbe1ad6439731b4312fb36813d6d/common/djangoapps/edxmako/middleware.py#L43-L44 - in templates it's done with `request.is_secure()` and `util.request.safe_get_host()`

Kelketek replied:
@antoviaque This is a management command. No request object is available.

antoviaque replied:
Ah, that makes sense - too bad :/ Maybe at least a configuration option to allow to set this, then.

Kelketek replied:
I could do that. It still begs the question of whether this is the best way to get the domain, though. I don't suppose you have an opinion on that particular matter?

antoviaque replied:
@Kelketek Thanks for making me look at this, I had missed that part - I don't think we currently use the django sites, at least I didn't see it used anywhere. I've more seen `settings.SITE_NAME` used, besides the method I was pointing earlier when the request object is available. Might be worth double-checking on IRC that it's the recommended approach in this case.

Kelketek replied:
@antoviaque I asked around on IRC and no one seemed to have a clue what to do with it :/ I could ask again.

antoviaque replied:
Yup, and if nothing comes out of it, I think you can go for `SITE_NAME`, at least you wouldn't be alone in using it.

antoviaque replied:
@Kelketek This still needs to be addressed

smarnach commented on lms/djangoapps/certificates/models.py:
```diff
@@ -574,3 +579,51 @@ def get_config(cls):
+        unique_together = (('course_id', 'user'),)
+
+
+# pylint: disable=unused-argument
```
You can prefix an argument with `unused_` to disable this warning for a single argument only.

Kelketek replied:
@smarnach I've got two arguments unused here :) Would you prefer I prefix both of them to having the pylint marker?

smarnach replied:
@Kelketek I usually go with the prefixes, but I don't feel strongly what you should do.

smarnach commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,134 @@
+        """
+        response = requests.post(self.assertion_url(mode), headers=self.get_headers(), data={'email': user.email})
+        self.log_if_raised(response)
+        BadgeTracker(course_id=self.course_key, user=user, data=response.json()).save()
```
Won't this throw if the badge has already been created?  We check whether the badge already exists in the signal handler before creating the BadgeHandler, but at this point we can't be sure anymore that it still does not exist.

Kelketek replied:
@smarnach I suppose if someone ran the regeneration command quickly twice in a row, this could happen. I'll code a contingency for it.

smarnach commented on lms/djangoapps/certificates/models.py:
```diff
@@ -574,3 +579,51 @@ def get_config(cls):
         return json_data
+
+
+class BadgeTracker(models.Model):
```
I find the name `BadgeTracker` really confusing.  This model describes badges, not badge trackers, so it should be called `Badge`.  The reason you write something into the database always is that you want to keep track of it, but you still don't call the model keeping track of users `UserTracker`. :)

smarnach commented:
Looks good to me, :+1:.

Kelketek commented:
@martynjames Could you give this a look over? Also, what is your opinion on merging it now and letting it lay dormant vs creating a feature branch?

Kelketek commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,151 @@
+                'mode': mode,
+                'assertion': assertion.data
+            }
+        )
```
@pbaruah Analytics event added. Happy to hear your thoughts on this PR, especially concerning the questions posed in the conversation area.

Doesn't look like I'll have a previewable version of this up today. Ran into some snags when trying to build atop https://github.com/edx/edx-platform/pull/7877 , see the comments there.

martynjames commented on lms/djangoapps/certificates/models.py:
```diff
@@ -574,3 +579,51 @@ def get_config(cls):
+    handler.award(instance.user)
+
+# Create badges when certificates are saved.
+post_save.connect(create_badge, sender=GeneratedCertificate, dispatch_uid="generate_badge")
```
Comment: I'm interested in the difference between using `post_save` as here, and using it within a function decorator (like `@receiver(post_save, sender=GeneratedCertificate)` as used above)

Kelketek replied:
Ah, I didn't see that earlier in the module. There should be no difference. I'll do it that way.

martynjames commented on lms/djangoapps/certificates/tests/test_badge_handler.py:
```diff
@@ -0,0 +1,173 @@
+        """
+        return BadgeHandler(self.course.location.course_key)
+
+    @override_settings(**BADGR_SETTINGS)
```
Optional: I believe that it is safe to override_settings for the whole class and they should apply for each test

Kelketek replied:
Turns out you're right. I'll push a change for this.

martynjames commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,151 @@
+        course = modulestore().get_course(self.course_key)
+        if mode not in settings.BADGR_IMAGE_SOURCES:
+            raise KeyError("No image specified for course mode {} in BADGR_IMAGE_SOURCES!".format(mode))
+        image_name = settings.BADGR_IMAGE_SOURCES[mode]
```
Question: So, there's only one 'honor' badge - the _same_ for all courses? Might be fine for now, but there's probably a desire to have different courses have different images - e.g. courses on chemistry might have an image with a breaker and testtube?

Kelketek replied:
Yes. The MVP specs state that, for now, we will have one image for each mode over all courses.

martynjames commented:
@Kelketek - Code looks good - I had a couple of questions, but really they are more for my understanding than criticism...

Is there a sandbox? Or perhaps a BADGR server available to point local code to?

Kelketek commented:
@martynjames I sort of have a sandbox, but there's not yet a UI to really allow one to see what's going on from that side. The command line would have to be used (and would have to be, even after a UI was made, to generate the badges). I'll answer your questions in a short bit here. What is your opinion on putting this into a feature branch vs merging it and letting it lay dormant for now? (Looks like I need to rerun Jenkins either way)

antoviaque commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -662,7 +662,14 @@ class CourseFields(object):
+            "Issue OpenBadges badges for this course. Badges are generated when certificates are created."
+        ),
+        scope=Scope.settings,
+        default=False
```
We talked of keeping this option as a way to deactivate it for some courses in the specs, but to put it on by default.

Kelketek replied:
I could turn it on by default, but that will mean some courses will have old versions of badges floating around that may not fit with their aesthetics. Some may want to wait until customization is available.

antoviaque replied:
Can you raise this issue in the specs? Since this was discussed there, I'd rather keep the discussion in one place. This should be true by default imho, that's mainly a platform-wide decision - most course authors won't even be aware of badges, this would severely limit their availability.

antoviaque replied:
@Kelketek This still needs to be addressed

antoviaque commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,151 @@
+        """
+        Base URL for all API requests.
+        """
+        return "{}/v1/issuer/issuers/{}".format(settings.BADGR_BASE_URL, settings.BADGR_ISSUER_SLUG)
```
This would add a double `/` if `BADGR_BASE_URL` contains a `/` at the end. Could be worth using https://docs.python.org/2/library/urlparse.html#urlparse.urljoin which is made for this.

Kelketek replied:
This is nice-- I didn't know about it. I'll use that.

Kelketek replied:
@antoviaque I'm finding this library's behavior difficult to reason through (it seems perfectly happy to chop off sections of the base URL) and it seems to make my job much harder here. I'm going to leave this as is, and add a note in the settings file to not include a trailing slash.

ottonomy replied:
@antoviaque @kelketek I haven't had a chance to look at all this code, but if you want to keep an eye toward issuing badges in overlapping contexts, I think it might be better to be more flexible  in this setting. I'd eventually hope edX would be interested in using API calls outside of `../issuer/`. Cheers!

antoviaque commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,151 @@
+            'slug': self.course_slug(mode)
+        }
+        result = requests.post(self.badge_create_url, headers=self.get_headers(), data=data, files=files)
+        self.log_if_raised(result)
```
When this happens, it would be good to log more than the response - also including the request could be helpful to debug and/or rerun the failed attempt.

Kelketek replied:
It would, but is there a way to log it without logging the API key? Or would we worry about that?

antoviaque replied:
The logs would only be accessible by admins, so I don't think including the API key would be an issue.

antoviaque replied:
@Kelketek This still needs to be addressed.

Also, if for some reason the certificate generation job is stopped mid-way, would an admin be able to resume it by rerunning it again?

Kelketek replied:
@antoviaque It should be fixed in my latest push. As for resuming the task, yes. The badges are generated when the certificates are set in the 'downloadable' status. So as long as the certificates end up in that state, the badges will be generated. Resuming the task would allow that to happen.

antoviaque replied:
Great, perfect!

antoviaque commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,151 @@
+        """
+        Slug ought to be deterministic and limited in size so it's not too big for Badgr.
+        """
+        return hashlib.sha256(u"{}{}".format(unicode(self.course_key), unicode(mode))).hexdigest()
```
It's also preferably human-readable - why use a hash here, and not a slugified course_key?

Kelketek replied:
Suppose a course name was all non-ascii unicode characters. Then it would have a high probability of collision.

antoviaque replied:
You can always include a random part at the end if you want - but it would be nice to be able to tell what the slug is about in most cases

antoviaque replied:
@Kelketek This still needs to be addressed

antoviaque commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,151 @@
+            domain = u'example.com'
+        about_path = reverse('about_course', kwargs={'course_id': unicode(self.course_key)})
+        data = {
+            'name': course.display_name,
```
It would be good to also include the course id, for cases where the display name would be ambiguous, since they are not guaranteed to be unique. Maybe in the description field?

Kelketek replied:
The course ID will be in the criteria URL below. Are you asking it to be added explicitly as well?

antoviaque replied:
Also, since the badges are per-mode, the course mode should likely also be included in the name of the badge.

Kelketek replied:
How would we get the translation in that case? There's no request object-- is there a course setting for language and a way to get a specific translation for that language? It sounds like a messy affair.

antoviaque replied:
What do you need the request object for in that case? To get the user preference, or to translate the string? I would assume that we don't need the request object for either of these, no?

antoviaque replied:
And yes, for the course_id I was suggesting to include it separately - but you're right, we can always reconstruct it from the criteria URL, so I don't feel too strongly about it.

Kelketek replied:
@antoviaque Well, yes, that's what we'd need the request for, but even that makes little sense. The main point I'm trying to make is that 'honor' or 'verified' is going to end up in a different language from whatever the course name is some of the time.

antoviaque replied:
@Kelketek Is there any translation in certificates? This looks like an issue that would appear in other places when generating certificates. Otherwise we could simply use the platform default language for now - since this would be at the course level, unless a course had a way to specify its language, there wouldn't be a way to tell which language to use here, even outside of the certificate generation?

Kelketek replied:
@antoviaque For the PDF certificates, no-- there's no way to figure out the language ahead of time and it's a bit of a mess dealing with it. You have to make your own custom rendering templates and whatnot. For the web certificates, we have access to the translation framework.

It should be noted that verified, honor, etc, are not only going to have separate badge images but will be separate badges. I guess the question is why we want this extra information in the assertion. If we put something in the name, it will show up as part of the name in whatever badge display service the user uses. This could look weird downstream.

antoviaque replied:
Here we are in the badge creation, not the assertion, right? These extra fields are for the badge yes. Currently they aren't included in the name, no?

antoviaque replied:
@Kelketek This still needs to be addressed - as a reminder, @pbaruah asked to enter the following elements: course title, course mode, course start & course end.

@pbaruah FYI the maximum size of the name field is 255 characters. Would the following format work for you?

```
*"[Course display name] ([Course mode], [date start] to [date end])"*
```

For example: *"Introduction to C++ (Honor, 2015-01-25 to 2015-04-30)"*

Also don't hesitate if I forgot any other information that should be included.

antoviaque commented on lms/envs/common.py:
```diff
@@ -1917,6 +1922,15 @@
+# A dictionary of course modes to file paths. Example:
+# {'honor': '/path/to/badge_image.png'}
+# Badge images must be square and in PNG or SVG format.
+BADGR_IMAGE_SOURCES = {}
```
Note that if you require image files to be present on the local filesystem, you would need to get the images there using ansible and create a PR against the configuration repo. It would likely be easier, both for you and for maintenance later on, to have this stored in a model, administrable via the django admin.

Kelketek replied:
@antoviaque What if there were default images that wouldn't need to be specially uploaded? Or if we used URLs instead, and the image would have to be downloaded? Since the download action would only have to be done upon the first badge generation (barring some slight race condition weirdness that shouldn't cause too much of a problem in practice), this might be an easier method.

antoviaque replied:
* Yes, it would be good to have defaults, so that it works better out of the box, without having to care about creating image badges initially.  It should be overridable though, as the default would likely be Open edX branded, while edx.org would have its own graphic I would imagine. @pbaruah Do we have image assets for the 3 badges already?
* Which raises the issue of updating the badge image when it is changed - especially the first time someone would try setting a badge, they might make a mistake - while we probably can't do much about badges that have already been generated, it would be tough to get stuck with the initial test image for all future badges generation. I don't see a command to update an existing badge in the API though (right?), so that would require recreating a new badge when it changes?
* Having a model makes it easier to have control over the file - besides ensuring that it is available at the time of the first request, it would also allow to check things like the size, or trigger a badge update when it is changed.

Kelketek replied:
@antoviaque I don't believe Badgr-Server has a method for updating a badge image. It's set in stone, at least in theory. However, I believe if one were particularly enterprising they could replace the image file in Badgr-server manually, and new badges would be created that way.

Kelketek replied:
Checking sizes and whatnot is going to be... different for SVG. I may have to add a library to the requirements just to do that. I like the idea of having the badgr server handle that validation.

antoviaque replied:
@ottonomy How would you go at changing the image of a badge with badgr? It doesn't seem possible with the current API - is this because it's not implemented yet, or are there other reasons?

ottonomy replied:
The current API doesn't let you update badges, and you can only delete them if they haven't been issued. If you make a mistake defining a badge, delete it before it has been issued and re-create it. 

As far as image storage for defined badges and issued (baked) badge images, the next upgrade to the Issuer server will allow you to use an AWS S3 bucket to store images, so they can be accessible across several application servers.

PNG support is all there is for now, but SVG will be coming this summer.

antoviaque replied:
@ottonomy Thanks for the replies!

@Kelketek See the comment about SVG support - if it isn't supported for now, we should probably remove the current mentions.

pbaruah commented:
@Kelketek - i) how are you going to add the two badge images (for honor and verified)? Till we have a configuration page in later phases where course authors will add badge images per course, we will have to put the images in a hard coded location for the MVP. Do you have any thoughts on this? 
ii) In addition to issuer, course name and student name for badge metadata, can we add some additional fields? I could think of course date and course mode (honor, verified). We can create an additional story for this if the effort is big - let me know.

antoviaque commented:
@pbaruah Cf the discussion above about the metadata fields: https://github.com/edx/edx-platform/pull/7888#discussion_r30127733

For course date and course mode, there are pending questions about translations, but provided that it's not too hard to add them, they could be included in the course name, to make it unique across course runs & modes.

Kelketek commented:
1. Give me your public SSH key ( @martynjames , I've imported yours already, if you'd like to go ahead and try)
2. ssh `edxapp@sandbox.opencraft.com`
3. `cd edx-platform;source ~/source ~/edxapp_env`
3. To delete all BadgeAssertion objects so new ones can be generated, do:

    ./manage.py lms shell --settings=aws
    from certificates.models import BadgeAssertion
    BadgeAssertion.objects.all().delete()

4. Drop out of the Python shell. To generate certificates (and thus generate a new badge):

    ./manage.py lms --settings aws regenerate_user -u honor@example.com -c edX/DemoX/Demo_Course --insecure

The honor user has passed the course and thus can have a certificate generated. You could also create courses and users to play with as you please, but this is the basic workflow.

mulby commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,151 @@
+                'user': user.id,
+                'course_id': unicode(self.course_key),
+                'mode': mode,
+                'assertion': assertion.data
```
Can this data vary in size? or is it a constant set of fields with well constrained data length?

Kelketek replied:
@mulby It won't vary much in size.

ottonomy replied:
@Kelketek is right that it will vary a little, but not much. There is an optional "evidence" field that may appear in some badges. Otherwise the set of fields is fairly constant. BadgeClasses have more variance with more optional fields and an unconstrained-length description field (though issuers are encouraged to keep their descriptions short)

mulby commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,151 @@
+            'edx.badges.assertion.created', {
+                'user': user.id,
+                'course_id': unicode(self.course_key),
+                'mode': mode,
```
I might call this "enrollment_mode" or put some other prefix indicating what "mode" refers to here.

mulby commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,151 @@
+        assertion, __ = BadgeAssertion.objects.get_or_create(course_id=self.course_key, user=user)
+        assertion.data = response.json()
+        assertion.save()
+        tracker.emit(
```
@lamagnifica - there is a new event here.

@Kelketek can you post a representative sample (or samples if there is expected to be some variation) in this PR so that @lamagnifica can see the shape of the events for documentation purposes?

Kelketek replied:
@lamagnifica @mulby Here's a sample event:

    {"username": "", "event_source": "server", "name": "edx.badges.assertion.created", "accept_language": "", "time": "2015-05-19T16:25:44.510759+00:00", "agent": "python-requests/0.14.2 CPython/2.7.3 Linux/3.2.0-80-virtual", "page": null, "host": "sandbox.opencraft.com", "session": "", "referer": "", "context": {"user_id": "", "org_id": "", "course_id": "", "path": "/update_certificate"}, "ip": "127.0.0.1", "event": {"course_id": "edX/DemoX/Demo_Course", "user": 1, "assertion": {"revocation_reason": null, "revoked": false, "created_at": "2015-05-19T16:25:44.379539Z", "created_by": "http://localhost:8005/user/staff", "email": "honor@example.com", "json": {"issuedOn": "2015-05-19T09:25:44.377599", "uid": "aa935ca3-b423-499a-b0f7-aed6e64bfcc9", "verify": {"url": "http://sandbox.opencraft.com:8005/public/assertions/aa935ca3-b423-499a-b0f7-aed6e64bfcc9", "type": "hosted"}, "image": "http://sandbox.opencraft.com:8005/public/assertions/aa935ca3-b423-499a-b0f7-aed6e64bfcc9/image", "recipient": {"type": "email", "salt": "3be29f13-944b-48bf-94ea-b0ab53661108", "hashed": true, "identity": "sha256$dd0e0451693821422203dca561acf06f50b13cf6db395aa012c9356fdd3049ab"}, "type": "Assertion", "@context": "https://w3id.org/openbadges/v1", "badge": "http://sandbox.opencraft.com:8005/public/badges/9b72927812da7558900762f6e06a163bf2adcdd712f34681dc6825a32fa20ef2", "id": "http://sandbox.opencraft.com:8005/public/assertions/aa935ca3-b423-499a-b0f7-aed6e64bfcc9"}, "badgeclass": "http://localhost:8005/public/badges/9b72927812da7558900762f6e06a163bf2adcdd712f34681dc6825a32fa20ef2", "image": "http://localhost:8005/media/issued/bc0d14ce0dbc36eeeef3dd7a27a919d2.png", "slug": "aa935ca3-b423-499a-b0f7-aed6e64bfcc9", "issuer": "http://localhost:8005/public/issuers/test-issuer"}, "mode": "honor"}, "event_type": "edx.badges.assertion.created"}

edx-webhook commented:
Thanks for the pull request, @Kelketek! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=7888

edx-webhook commented:
Thanks for the pull request, @Kelketek! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=7888

pbaruah commented:
@antoviaque - The UX team has the story in their current sprint to provide the two badge images - one for honor and one for verified. I see you mentioned three images - were you expecting one more? (prof ed?) cc: @marcotuts

antoviaque commented:
@pbaruah Yes, exactly - based on the description of https://openedx.atlassian.net/browse/SOL-768: 

> The badge image will be different for different course types like honor, verified and professional education courses.

Kelketek commented:
@pbaruah When will these images be ready, and is there a link to the mentioned ticket?

@antoviaque I've addressed most of the notes in my local branch but am in the middle of implementing the configuration structure. Here's the plan:

1. Have a base configuration in the `common.py` settings file that has file paths. These will point to the sample badges in one of the static folders of the repository directly, and load them from disk. @pbaruah , are we able to use the images the UX team is working on for that purpose?
2. Have a configuration model that accepts URLs pointing toward images to be used as the default. This will be the preferred override method. The model that handles this will do the lookup of the images via a method that will fallback to the settings paths if the mode is not defined in the configuration.
3. (Probably not to be implemented in the MVP), have a configuration dict on the course that points to assets the course author has uploaded. The configuration model above will accept the course key as an argument to its searching method, and will load from the contentstore an image if one is specified first.

antoviaque commented:
@Kelketek

* 1 & 2 are mutually exclusive - if you have a configuration model why would you need a configuration setting in common.py?
* If the model only contains the URL and not the image itself, how do you detect when the image changes?
* You might already have this in mind, but there can be an arbitrary number of modes, potentially - we should be doing something sensible with modes that aren't listed, and allow to potentially configure these additional modes without requiring a migration of the model.
* Yes, 3 isn't in the MVP specs, so it shouldn't be included for now, it will come later on

What about:
* One object per mode/image combination, including the image content in the model
* Using normal models rather than configuration models, which are made to be singletons
* Later on, implementing 3 by adding an optional course field to the model (objects without a course field remaining the default for all courses on a given mode). Also, a badge without either a course or a mode would be the default for unlisted modes

antoviaque commented:
@pbaruah @marcotuts Yes, are the badges being produced for Open edX or for edx.org? Ideally now we need ones that we can use as defaults on Open edX.

pbaruah commented:
@Kelketek , @antoviaque - The sample badges being created by the UX team are for Open edX only. It is covered in UX-2196 and in the current UX sprint, assigned to @talbs .

antoviaque commented:
@pbaruah Great, thanks for the details! I see that the story is currently in the backlog in the current sprint, which is perfect. @talbs Are the icons on track for completion this sprint? @Kelketek will need it to complete the current PR, so if you have an idea of when he will be able to integrate them in this PR, that would be useful. Thank you!

antoviaque commented on lms/djangoapps/certificates/models.py:
```diff
@@ -574,3 +579,92 @@ def get_config(cls):
+        # Actual max is 256KB, but need overhead for badge baking. This should be more than enough.
+        help_text=_(
+            u"Badge images must be square PNG files. To use with Mozilla Backpack, the filesize should be "
+            u"under 250 kilobytes."
```
We talked about validating this on the model, upon upload - since these are the generic images, you can't really create the badges on the badgr server now to fully validate the badge creation of a specific course, but you can check the image type and size. Another option would be to create a sample badge on the server, just for the purpose of validating the image, if you want to fully delegate this to badgr and avoid duplicating the checks.

antoviaque commented on lms/djangoapps/certificates/models.py:
```diff
@@ -574,3 +579,92 @@ def get_config(cls):
+            u"default image."
+        )
+    )
+
```
Can you add a data migration to include default badges? While we wait for UI to send the final badges, it can be mock images, this way we would just have to swap them when they are ready.

Kelketek replied:
@antoviaque This has been added.

talbs commented:
@antoviaque, thanks for the ping about the timing of these placeholder badge assets. I'm working on them now and expecting to have something within the next 24-36 hours. I will create a PR against this once I have some quick review/sign off on the assets.

Shout if you have any concerns.

Kelketek commented:
Jenkins, test this please.

Kelketek commented:
@talbs This branch now has some sample images you can override as soon as you're ready to make a PR for it.

Kelketek commented:
@pbaruah The course mode is now included in the name of created badges. We have a few issues in terms of limitations, however.

Badges will be generated with the default locale translation, and not with any particular language. Since the badges are generated on callback, no user is associated with them and so no special locale is loaded. The badge description currently just lists the date range, per @antoviaque 's suggestion to me. I'm not sure what, specifically, else to put there. If you have a particular suggestion, let me know.

As soon as @talbs provides the images, this should, in theory, be ready to merge, so long as @mulby has no more comments on the currently included analytics event. The rest of the events will be created in a separate PR.

@martynjames Please let me know if there's anything else I need to do to get your +1 on this. It didn't look like you had a whole lot of particular notes to begin with, but I want to make sure I'm in the clear.

I'm going to be working on the UI stuff to help @talbs and @mattdrayer along with asset compilation issues, and I'll make a temporary version of the UI code on top of that, which I'll be fleshing out a bit more now that their code is beginning to crystallize.

martynjames commented:
:+1: from me on this - I like the approach - lots of room to grow as we enhance the badging feature more.

antoviaque commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,175 @@
+            LOGGER.error(
+                u"Encountered an error when contacting the Badgr-Server. Request sent to %s with headers %s.\n"
+                u"Response status was %s.\n%s",
+                repr(response.request.url), repr(response.request.headers),
```
In the case of a POST request, that would be missing the request body (`response.request.body`), which contains the parameters and thus the bulk of the request content.

antoviaque commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,175 @@
+        about_path = reverse('about_course', kwargs={'course_id': unicode(self.course_key)})
+        data = {
+            'name': self.badge_name_field(course, mode),
+            'criteria': u'http://{}{}'.format(settings.SITE_NAME, about_path),
```
Still missing a configuration option to optionally use https URLs

antoviaque commented:
@talbs That's perfect, thank you!

antoviaque commented on lms/djangoapps/certificates/migrations/0023_default_modes.py:
```diff
@@ -0,0 +1,151 @@
+
+    def forwards(self, orm):
+        """Add default modes"""
+        for mode in ['honor', 'verified', 'professional']:
```
Would be good to mark one of the badges as the default one, to avoid failing badge creation for other modes - honor is the lowest certification level, so out of the three available now that would likely be the one. Could be good to replace it with a more generic one later on though -- cc @pbaruah

antoviaque commented:
@Kelketek Thanks for the changes! I had a few follow-up comments, but this LGTM now - :+1: provided that they are addressed.

mulby commented on lms/djangoapps/certificates/badge_handler.py:
```diff
@@ -0,0 +1,182 @@
+                'user': user.id,
+                'course_id': unicode(self.course_key),
+                'enrollment_mode': mode,
+                'assertion': assertion.data
```
After thinking about this more, I'm a little nervous about just dumping the response in here. Primarily because it means we don't control the data that is going into the event, which can cause problems:

1) The size of this payload increases dramatically.
2) The remote service changes the payload to include data that we do not want to save in our tracking logs.
3) The remote service changes the format and/or structure of the data which, in turn, breaks the code of researchers who are analyzing the log and making assumptions about the format and structure.

I think I would feel better about it if we explicitly copied the data we want from the response into the event.

mulby commented:
It looks like the event emission code here isn't complete yet, I may have jumped the gun on the review, sorry about that. I provided some preliminary feedback here in the review and on the proposal wiki page (thanks for creating that!).

Also, a general note, please add tests for the event emission code, I wrote up some docs here: http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/analytics.html#testing-event-emission

It should be fairly straightforward to add tests to the unit test suite, the bok choy suite or both.

Ping me when this is ready for analytics review again.

Kelketek commented:
@mulby Only the single event in this PR will be included with it. I will be making a follow-up PR for the others. I'll look at updating this event per your notes.

Kelketek commented:
@mulby I've updated the event, and added an assertion for it in the tests. The Analytics doc has been updated to reflect the changes. Let me know if these fields will do for you.

Kelketek commented:
Also, here is a new example of the event, @lamagnifica @mulby 

```
{
  "username": "",
  "event_source": "server",
  "name": "edx.badges.assertion.created",
  "accept_language": "",
  "time": "2015-05-21T19:20:43.338774+00:00",
  "agent": "python-requests/0.14.2 CPython/2.7.3 Linux/3.2.0-23-generic",
  "page": null,
  "host": "localhost",
  "session": "",
  "referer": "",
  "context": {
    "user_id": "",
    "org_id": "",
    "course_id": "",
    "path": "/update_certificate"
  },
  "ip": "127.0.0.1",
  "event": {
    "assertion_slug": "c44ba624-490a-4b0e-8a47-de072c1a2bdd",
    "enrollment_mode": "honor",
    "user": 1,
    "assertion_json_url": "http://example.com/public/assertions/c44ba624-490a-4b0e-8a47-de072c1a2bdd",
    "course_id": "course-v1:edX+DemoX+Demo_Course",
    "issuer": "http://example.com/public/issuers/test-issuer",
    "badge_slug": "course-v1edxdemoxdemo_course_honor_924c2c1",
    "assertion_image_url": "http://example.com/media/issued/dfc31731b946de5bd23d92a100c9d05f.png"
  },
  "event_type": "edx.badges.assertion.created"
}
```

lamagnifica commented:
Excellent thank you! @catong ^

Kelketek commented:
@lamagnifica @catong @mulby I'm going to make one tweak now that I'm looking at the event pretty-printed. I'm going to change `user` to `user_id`.

talbs commented:
@Kelketek and @antoviaque, I'm attaching two placeholder badge images (based on the open certificate states of base and distinguished). I didn't want to make a PR against this as I'm not sure the file names you're looking for match the cases I designed for (again I'm sticking with my ``*-base`` or ``*-distingushed`` file naming for these). 

In the next few days, I'll be making a separate PR against ``mattdrayer/SOL-465`` to sync up the certificate rendering assets (distinguished's icon + watermark background) visually with these sample badges.

Let me know if these course-centric badges don't fit the bill or if you folks need anything else.

- - -

![badge-base](https://cloud.githubusercontent.com/assets/163763/7763440/aae6682a-000e-11e5-81fe-6a9868249300.png)
![badge-distingushed](https://cloud.githubusercontent.com/assets/163763/7763439/aae5ab1a-000e-11e5-8bd4-8a6fddec5625.png)

pbaruah commented:
@Kelketek , @antoviaque - things are coming along nicely. Thanks!

Locale: If an open edx installation for e.g. in Saudi Arabia sets up the badge issuing service, the language will be in Arabic, which is fine. Later on, we might have to find out if specific language will have to be set, I don't know yet.

Format: *"[Course display name] ([Course mode], [date start] to [date end])"*
This is fine. The 255 char limitation should be enough for our course titles. Please make sure that for a self-paced course, there is no start and end date.

Default course mode: You mentioned that we should set honor as the default in case badge generation for the other modes fail. I think it is better to generate an error and not issue a badge than to generate the wrong badge (i.e. a verified student will not get a honor badge). The error would require a regeneration of badges.

Good points in the discussion on analytics events. I will also respond to the discussion there. Would it be possible to generate a sample badge against my email id so that I can upload to backpack and see how it looks?

@talbs - The badge images look great.

ottonomy commented:
@pbaruah @Kelketek @antoviaque: Please let me know if you hear of any difficulties using the Badgr software in Open edX instances that aren't using English. We haven't done any translation of the software or API yet, but we hope to be able to do proper internationalization eventually. (And within the Open Badges specification, we have begun to adopt some technologies that will make it far more possible to support translation of badge resources).

Kelketek commented:
@mulby @lamagnifica @catong I have removed the analytics event that was in this PR, and will be creating a new PR shortly that focuses only on analytics events for the badging feature.

@talbs Thanks for the images! I've put them in the PR where they should be. @pbaruah , since there are only two images, I've used the more flashy one for verified and professional. I'm not sure if that's the right way to go, but it should do for now, I think.

@pbaruah I have removed the automatic setting of a default. It should fail if they try to generate with a mode that doesn't exist now.

All: I will be closing this PR, rebasing the branch, and then pushing it as a feature branch. I will then re-open a PR that will act as the 'MVP PR', and open another PR that will be for Analytics atop this new feature branch. When Matt's work is finished, the UI PR will be laid on top of this as well, allowing for a final review on the MVP branch before it is merged in.

catong commented:
@Kelketek I appreciate the heads-up. Please tag me on the new MVP and Analytics PRs -- thanks!

pbaruah commented:
@Kelketek - yes, using a copy of the "more prominent" image for professional courses should be OK for now. I'll see if we can get another image for professional in which case we will replace it before the release.
